### ISSUE TITLE
Some questions regarding architecture

### ISSUE THREAD

**Issue by samin9796** [facebookresearch/wav2letter#763]
1. Could you please explain the parameters of the following layer and what it does?

`RO [firstDim] [secondDim] [thirdDim] [fourthDim]`

2. Is it possible to add FC layers or RNN layers after Convolutional layers using wav2letter++? 

3. When I train with 20 conv layers, I get the best result. But when I try out 25 conv layers, WER is more than 100%. Why does accuracy drop so drastically? My arch is like conv layer followed by relu activation function and then dropout layer with probability 0.1. 

4. Could you tell me a bit about LM smearing? I found max smearing improves accuracy significantly but can't find what it actually is.

Thank you.

**Comment by tlikhomanenko** [comment]
> 1. Could you please explain the parameters of the following layer and what it does?

RO - it is reordering  of the dimensions. For example if you have tensor [T, 1, C, B] and you want to have [C, 1, T, B] then you should use `RO 2 1 0 3`

> * Is it possible to add FC layers or RNN layers after Convolutional layers using wav2letter++?

Yep, sure, check for example here how we define different architectures https://github.com/facebookresearch/wav2letter/tree/master/recipes/models/sota/2019/am_arch, just have in mind you need to have dimensions in agreement.

> 2\. When I train with 20 conv layers, I get the best result. But when I try out 25 conv layers, WER is more than 100%. Why does accuracy drop so drastically? My arch is like conv layer followed by relu activation function and then dropout layer with probability 0.1.

Just increasing number of layers could not work with the same parameters, possibly you need to tweak learning rate and optimization at least.

> 3\. Could you tell me a bit about LM smearing? I found max smearing improves accuracy significantly but can't find what it actually is.

In case of word LM when we add LM score only when we reach terminal node (=we form a word) we are doing some heuristic and start to add LM score early as we proceed with tokens. To do so we propagate unigram LM score for the word from terminal node to the root, so thus we are adding some score at each token, however, finally we will add the LM score of the word in current context. While propagate score in the trie we can do max of children scores (corresponds to the smearing=max) or logadd operation.

**Comment by samin9796** [comment]
@tlikhomanenko  
Thank you. In the wiki section (Writing architecture files), I can't find any information about FC layers. How can I add some FC layers?

**Comment by samin9796** [comment]
`LSM [normDim]`
`PR [numElements <OPTIONAL>] [initValue <OPTIONAL>]`

What is normDim here? And what can be a good value of normDim? Could you provide any information about numElements and initValue?

**Comment by tlikhomanenko** [comment]
@samin9796 

FC layer is just linear layer:
```
fl::Linear
L [inputChannels] [outputChannels]
```

`LSM [normDim]` - norm dim is dimension along which log softmax will be done. Like if you normalize across first dim then jsut set `LSM 0` (as c++ style starting numeration from 0 :) )

On PReLU here is doc in flashlight https://github.com/facebookresearch/flashlight/blob/716d1b0913be1c26e446ed37450a5081c371a749/flashlight/nn/modules/Activations.h#L170, so the initValue is coeff for the linear function for x < 0. Default is 0.25, so you can start with it (as in original paper on PReLU https://arxiv.org/pdf/1502.01852.pdf), numElements is the number of learnable coeff of PReLU - it should be a multiple of first dim of input tensor. You can simply set it to 1 so the same one learnable coeff will be learned during training for PReLU and the same activation will be applied for any element of input tensor (feel free to check implementation https://github.com/facebookresearch/flashlight/blob/716d1b0913be1c26e446ed37450a5081c371a749/flashlight/nn/modules/Activations.cpp#L96).
